Cps1 bootlegs, some more new sets

TwistedTom opened the pull request:
Some new sets in Fcrash:
Added captcommb2 - Captain Commando single board bootleg (w/ 2x oki MSM5205)
Added knightsb3 - Knights of the Round single board bootleg (w/ 2x oki MSM5205)
Added dinopic3 - Cadillacs and Dinosaurs single board bootleg (w/ PIC) no sound (pic secured)

TwistedTom commented:
Indeed the cps1 bootlegs are a bit of a mess, some in cps1.cpp, some in fcrash.cpp. I think it would be good to remove all the bootlegs out of cps1, keep that clean and for originals only. Would you prefer all bootlegs in a single source file eg. cps1bl.cpp or lots of per game files eg. fcrash? But then fcrash has lots other games in it already so they would need taking out.

cuavas commented:
I'm not asking you to clean it all up right now, just to add a derived state class for the game you're adding.  If you want to split other games into derived classes while you're at it that's great, but it's not a blocker for this PR.

As for the ideal state, eventually we'd probably want to have originals and possibly pure ROM swap hacks/bootlegs in cps1.cpp, and bootlegs moved to other files grouped by related hardware.  If fcrash.cpp is a collection of similar bootlegs, they can stay together.

TwistedTom commented:
OK, I think i'll create 2 derived state classes for "ym2151+2xmsm5202" bootleg h/w and "PIC-for-sound" bootleg h/w. Put relevant existing sets and my new dumps in those. That should clean things up a fair bit for now?

cuavas commented:
Sounds good.

TwistedTom commented:
So i've done bit of work on this, seems to me there's 5 catergories:
1. Rom swaps, hacks etc.  (on original Capcom hardware)  ->  cps1.cpp
2. Sound: Z80, 2x YM2203, 2x m5205 ("Final Crash" h/w)   ->  fcrash.cpp
3. Sound: Z80, 1x YM2151, 2x m5205                       ->  new derived class
4. Sound: PIC, 1x M6295                                  ->  new derived class
5. Sound: Z80, 1x YM2151, 1x M6295 *                     ->  still in fcrash.cpp for now...
(5. this is original configuration, but non-Capcom (usually single-board) hardware.)

so far i've...
made fcrash derived class of cps_state
moved fcrash specific stuff out of cps_state
made new cps1bl_5205 derived class + source file
made new cps1bl_pic derived class + source file
moved relevant games out of fcrash into new classes
general tidy-up of cps1.h
general tidy-up of fcrash.cpp
improved sound for existing "YM2151+2xMSM5205" games
added new dumps as per original PR

I think that's tidied things up quite a bit for now?

Work still to do is:
make a new class/source file for #5
move non #1 games out of cps1.cpp (probably most of them)
move non #2 games out of fcrash
remove bootleg specific code from cps1.cpp (kludges, "reverse" sprite rendering etc.)
general tidy-up/modernisation of cps1.cpp

cuavas commented on src/mame/includes/cps1.h:
```diff
@@ -525,7 +408,184 @@ class cps2_state : public cps_state
+	optional_device<msm5205_device> m_msm_1;
+	optional_device<msm5205_device> m_msm_2;
+	
+	void (fcrash_state::*m_bootleg_sprite_renderer)(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
```
Would this be better to do with a virtual member function?

TwistedTom replied:
That was troubling me a bit as well! A virtual would work nice if every game in the derived class used the same function, but unfortunately they don't. I need a way to swap the function if a game needs it (in game_init), if not just do nothing and the base version is used. I'm sure there's nicer oo way but it's eluding me!

cuavas replied:
How many different implementations of that function are there?  You only need one derived class per implementation, and you can have multiple derived classes in a source file (e.g. see mw8080bw.cpp where there are a pile of derived classes for different hardware variations).

TwistedTom replied:
There's just one game in each derived class using specialised function and only one implementation of it. If a per-game derived class is ok then that's fine with me (I thought that perhaps overkill just for one function). I will do that.
